### ISSUE TITLE
Cannot sign with in-memory key '[...]:signMavenPublication' because it has no configured signatory

### ISSUE THREAD

**Issue by victorpaleologue** [vanniktech/gradle-maven-publish-plugin#248]
As explained in the README: https://github.com/vanniktech/gradle-maven-publish-plugin#signing
I moved to in-memory GPG keys, using the following Gradle properties:
```ini
signingInMemoryKey=...
signingInMemoryKeyId=...
signingInMemoryPassword=...
```
I used `gpg --export-secret-keys --armor $KEY_ID` and assigned the data to `signingInMemoryKey`.
I also tried with the base64 version.
It really seems the key is well decrypted, because when I did it wrong I had the error `Could not read PGP secret key`, and I do not have it here.

I did not change my gradle code, but now it fails with the message: `Cannot perform signing task '[...]:signMavenPublication' because it has no configured signatory`
You can see the gradle code here: https://github.com/aldebaran/pddl-planning-android/blob/a97d0656b7fe65cc0668b37c9e0da89e044932f0/pddl-planning/build.gradle

And the full error report with the stacktrace here: https://pastebin.com/3aiRjsZz

I am probably doing something wrong, but I cannot find any clue about it in the README.

**Comment by victorpaleologue** [comment]
I am not sure I understand Gradle well, but here is the first serious clue I found.

Looking at the plug-in's code, the error message corresponds to the absence of a `Signatory` object. It seems that calling `useInMemoryPgpKeys` [is meant to set signatories](https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/signing/src/main/java/org/gradle/plugins/signing/SigningExtension.java#L248).

In [Gradle's `signing` plug-in tests](https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/signing/src/integTest/groovy/org/gradle/plugins/signing/InMemoryPgpSignatoryProviderIntegrationSpec.groovy), the only signing declaration to be tested puts the call to `useInMemoryPgpKeys` before the call to `sign`.

But in `gradle-maven-publish-plugin`, [`sign` is called before `useInMemoryPgpKeys`](https://github.com/vanniktech/gradle-maven-publish-plugin/blob/7d9a7b15dff77d67c5d782a4ae8664c53459cc2a/src/main/kotlin/com/vanniktech/maven/publish/MavenPublishBaseExtension.kt#L131). I understand that Gradle framework is rather declarative, so the order of calls might not be relevant. But it is not impossible that order matters because of implementation details in Gradle's `signing` plug-in.

**Comment by a914-gowtham** [comment]
@victorpaleologue  check this issue #71

**Comment by gabrielittner** [comment]
@victorpaleologue I can reproduce your issue. Could you try removing `signingInMemoryKey` from gradle.properties (the other 2 can stay) and create an environment variable like this `export ORG_GRADLE_PROJECT_signingInMemoryKey=...`?

**Comment by a914-gowtham** [comment]
@victorpaleologue  These three values are `signingInMemoryKey`, `signingInMemoryKeyId` and `signingInMemoryPassword` to be placed in same ~/.gradle/gradle.properties file?

**Comment by victorpaleologue** [comment]
> @victorpaleologue I can reproduce your issue. Could you try removing `signingInMemoryKey` from gradle.properties (the other 2 can stay) and create an environment variable like this `export ORG_GRADLE_PROJECT_signingInMemoryKey=...`?

I have just tried and it produces exactly the same output.

> @victorpaleologue These `signingInMemoryKey`, `signingInMemoryKeyId` and `signingInMemoryPassword` values are to be placed in same ~/.gradle/gradle.properties file?

Yes.

**Comment by victorpaleologue** [comment]
> @victorpaleologue check this issue #71

Right, it is in fact the same issue. But I cannot use encrypted key files because I am trying to build from GitHub actions, and GitHub secrets only support plain strings.

**Comment by twyatt** [comment]
I encountered this failure when I had one or more of my `singingInMemory*` properties configured incorrectly.

At first, my `singingInMemoryKey` **wasn't** stripped of the header/footer (as described in https://github.com/vanniktech/gradle-maven-publish-plugin/pull/201#discussion_r584270633).

Then, I had specified the wrong `signingInMemoryKeyId`. I ultimately ended up just removing this property, as it is optional.

I also had specified the password under `signingInMemoryPassword` before realizing that I was missing the `Key` portion of the property name (i.e. it is supposed to be `signingInMemoryKeyPassword` — [`MavenPublishBaseExtension.kt#L121`](https://github.com/vanniktech/gradle-maven-publish-plugin/blame/4b3df71b859bf5260d90542a12c991ed31d42af2/src/main/kotlin/com/vanniktech/maven/publish/MavenPublishBaseExtension.kt#L121)).
